packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1251 @@
+import { Notifier, Subscriber, Observable } from '@microsoft/fast-element';
+import {
+    keyArrowDown,
+    keyArrowLeft,
+    keyArrowRight,
+    keyArrowUp,
+    keyEnd,
+    keyEnter,
+    keyEscape,
+    keyFunction2,
+    keyHome,
+    keyPageDown,
+    keyPageUp,
+    keySpace,
+    keyTab
+} from '@microsoft/fast-web-utilities';
+import { FoundationElement } from '@microsoft/fast-foundation';
+import type { ScrollToOptions } from '@tanstack/virtual-core';
+import type { Table } from '..';
+import {
+    TableActionMenuToggleEventDetail,
+    TableFocusType,
+    TableHeaderFocusableElements,
+    TableRowFocusableElements,
+    type TableRecord
+} from '../types';
+import type { Virtualizer } from './virtualizer';
+import { TableGroupRow } from '../components/group-row';
+import { TableRow } from '../components/row';
+import { TableCell } from '../components/cell';
+import { MenuButton } from '../../menu-button';
+import { tableHeaderTag } from '../components/header';
+import { TableCellView } from '../../table-column/base/cell-view';
+
+interface TableFocusState {
+    focusType: TableFocusType;
+    headerActionIndex?: number;
+    rowIndex?: number;
+    columnIndex?: number;
+    cellContentIndex?: number;
+}
+
+/**
+ * Manages the keyboard navigation and focus within the table.
+ * @internal
+ */
+export class KeyboardNavigationManager<TData extends TableRecord>
+implements Subscriber {
+    private focusType: TableFocusType = TableFocusType.none;
```

Comment:
In the [draft PR](https://github.com/ni/nimble/pull/2070), there was some discussion on splitting out the properties dealing with focus (`focusType` and `row/column/headerAction/cellContentIndex`) to a separate class, and also moving the methods that transition between focus states to that same class. I experimented with that in earlier drafts, but didn't feel like it really made the code any more readable/shorter. If you want to see a partial implementation of that, I published that earlier branch here: [separated TableFocusState](https://github.com/ni/nimble/blob/users/masmith/table-kbd-nav-temp-0524-separateFocusState/packages/nimble-components/src/table/models/table-focus-state.ts), [KeyboardNavigationManager](https://github.com/ni/nimble/blob/users/masmith/table-kbd-nav-temp-0524-separateFocusState/packages/nimble-components/src/table/models/keyboard-navigation-manager.ts). One issue was that I still need to update some additional state of this class (like `inNavigationMode`).

Given that, I decided to 'flatten' `focusType` and the `index` properties into this class, as it shortened their usages (`this.focusState.focusType` => `this.focusType`). And currently `TableFocusState` is only used where I need to encapsulate all of the focus state in the Tab/Shift-Tab handling code.

As usual I'm open to revisiting this if reviewers think that there's a cleaner way to write this.